sunpy/io/cdf.py

```diff
@@ -71,18 +71,28 @@ def read_cdf(fname):
 
             data = cdf.varget(var_key)
             # Get units
-            unit_str = attrs['UNITS']
             try:
-                unit = u.Unit(unit_str)
-            except ValueError:
-                if unit_str in _known_units:
-                    unit = _known_units[unit_str]
-                else:
-                    warn_user(f'astropy did not recognize units of "{unit_str}". '
+                unit_str = attrs['UNITS']
+                try:
+                    unit = u.Unit(unit_str)
+                except ValueError:
+                    if unit_str in _known_units:
+                        unit = _known_units[unit_str]
+                    else:
+                        warn_user(f'astropy did not recognize units of "{unit_str}". '
+                                  'Assigning dimensionless units. '
+                                  'If you think this unit should not be dimensionless, '
+                                  'please raise an issue at https://github.com/sunpy/sunpy/issues')
+                        unit = u.dimensionless_unscaled
+            except KeyError as keyerr:
+                if keyerr.args[0] == 'UNITS':
+                    warn_user(f'No units provided for variable "{var_key}". '
                               'Assigning dimensionless units. '
-                              'If you think this unit should not be dimensionless, '
+                              'If you think this variable should not be without units, '
```

Comment:
In the message I don't think we should tell users to raise an issue, as we don't want to be maintaining a list of all variables names out there that don't have a units field.

Reply:
Was thinking about that, too. Reducing it to `'No units provided for variable "{var_key}". Assigning dimensionless units.'`